imx93: add rgpio driver support

CkovMk opened the pull request:
Add RGPIO driver and enable it on i.MX 93.
Update clock_mcux_ccm_rev2. Change i.MX93 to use ccm_rev2 driver.

The HAL must be updated using https://github.com/zephyrproject-rtos/hal_nxp/pull/273  and https://github.com/zephyrproject-rtos/hal_nxp/pull/278.

danieldegrasse commented:
Hi @CkovMk- is this PR dependent on #62755? Also, please update the `west.yml` file to point to the PR that is needed for this support, using the `pull/PR_NUM/head` syntax: https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules

CkovMk commented:
Hi @danieldegrasse , `west.yam` has been updated. Thanks

mmahadevan108 commented:
@CkovMk , kindly update `west.yml` and help address other CI issues.

CkovMk committed:
board: mimx93_evk_a55: add led and button definitions and alias

Tested with zephyr example project "samples/basic/blinky" and "samples/
basic/button". These examples can run out-of-the-box. No modification
needed.
for "samples/basic/blinky", the red LED inside RGB LED on board will
blink every 2 seconds.
for "samples/basic/button", the red LED inside RGB LED on board will
turn on once BTN1 on board is pressed down. An log will be present in
uart console as well.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
@mmahadevan108 ci issues resolved. compliance check reports non-issue.

I'm not sure if gpio related samples/tests are run. Please double check.

CkovMk commented:
trying to add gpio test

CkovMk commented:
rebase on main for west.yml

mmahadevan108 commented:
@danieldegrasse @decsny , can you please re-review

danieldegrasse commented on drivers/gpio/Kconfig.mcux_rgpio:
```diff
@@ -0,0 +1,12 @@
+# MCUX RGPIO configuration options
+
+# Copyright (c) 2023, NXP
```
We are generally trying to format copyright statements as `Copyright 2023, NXP`. Is this format OK with you?

CkovMk replied:
replaced with new statement.

danieldegrasse commented on drivers/gpio/gpio_mcux_rgpio.c:
```diff
@@ -0,0 +1,329 @@
+	RGPIO_Type *base = (RGPIO_Type *)DEVICE_MMIO_NAMED_GET(dev, reg_base);
+	const struct mcux_rgpio_config *config = dev->config;
+
+#ifdef CONFIG_PINCTRL
```
I'm not sure we need the `CONFIG_PINCTRL` dependencies anymore. These were generally removed, see #57042

CkovMk replied:
removed.

danieldegrasse commented on soc/arm64/nxp_imx/mimx9/pinctrl_soc.h:
```diff
@@ -19,7 +19,6 @@ extern "C" {
 #define MCUX_IMX_DRIVE_OPEN_DRAIN_SHIFT IOMUXC1_SW_PAD_CTL_PAD_OD_SHIFT
 #define MCUX_IMX_BIAS_PULL_DOWN_SHIFT IOMUXC1_SW_PAD_CTL_PAD_PD_SHIFT
 #define MCUX_IMX_BIAS_PULL_UP_SHIFT IOMUXC1_SW_PAD_CTL_PAD_PU_SHIFT
-#define MCUX_IMX_BIAS_PULL_ENABLE_SHIFT IOMUXC1_SW_PAD_CTL_PAD_PE_SHIFT
```
Just checking- is this field being dropped because it is not present on the hardware? I don't see a `PE` field in the IOMUXC registers for the iMX93, so I assume so.

CkovMk replied:
Correct. i.MX93 only uses PU and PD field.

CkovMk commented on drivers/gpio/Kconfig.mcux_rgpio:
```diff
@@ -0,0 +1,12 @@
+config GPIO_MCUX_RGPIO
+	bool "MCUX RGPIO driver"
+	default y
+	depends on HAS_MCUX_RGPIO
```
removed HAS_MCUX...

CkovMk commented:
Hi @danieldegrasse ,

After removing the HAS_MCUX_RGPIO, it seems we got some issue here. There are already 2 drivers claims compatiable to "nxp,imx-gpio", namely `gpio_imx.c` and `gpio_mcux_igpio.c`.  It seems they relies on HAS_MCUX... to distinguish which one to use. This RGPIO driver follows the same approach and use the same compatiable string. Since now HAS_MCUX_RGPIO is removed, CI build shows error from targets that should NOT build this driver in the first place.

I surely can change this driver to claim compatiable to something like "nxp,imx-rgpio". Should we also rename the IGPIO driver?

danieldegrasse commented:
> I surely can change this driver to claim compatiable to something like "nxp,imx-rgpio". Should we also rename the IGPIO driver?

Thanks for pointing this out- I do think we should look at how to correct this. It looks like `gpio_imx.c` and `gpio_mcux_igpio.c` both manage the same hardware IP block, so we should not change the compatible. Since the rgpio driver is a new IP block, it should have a new compatible.

I can look at addressing the issue with the `gpio_imx.c` and `gpio_mcux_igpio.c` drivers. For the RGPIO driver, will using a new compatible string resolve this issue?

CkovMk committed:
driver: add new gpio driver "gpio_mcux_rgpio"

Add RGPIO gpio driver. This driver is used for i.MX93 and i.MX8ULP.
GPIO pinctrl, read/write and interrupt is supported. Runtime mmio
configuration is enabled, so no need for region definition in
mimx9/mmu_region.c

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
> will using a new compatible string resolve this issue?

A new compatiable string "nxp,imx-rgpio" created. Build and tested OK.

CkovMk committed:
board: mimx93_evk_a55: enable rgpio

Add RGPIO 1,2,3,4 to mimx93_evk_a55 board dts

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

henrikbrixandersen commented:
@CkovMk Can you confirm that the GPIO testsuites under `tests/drivers/gpio/` all pass on hardware with this driver?

CkovMk committed:
soc: imx93: enable rgpio driver

Add HAS_MCUX_RGPIO to Kconfig.soc
Add gpio1 ~ gpio4 dts node and pinctrl node in nxp_mimx93_a55.dtsi

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk committed:
tests: drivers: gpio_basic_api add board mimx93_evk_a55

Add mimx93_evk_a55.overlay to tests: drivers: gpio_basic_api

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
@henrikbrixandersen I ran gpio_basic_test and it passed. I did not run others, since I notice there aren't many boards support them. We've been using this driver in several demo projects, includes input/output and interrupt without issue. the button sample also works fine.

henrikbrixandersen commented:
> @henrikbrixandersen I ran gpio_basic_test and it passed. I did not run others, since I notice there aren't many boards support them. We've been using this driver in several demo projects, includes input/output and interrupt without issue. the button sample also works fine.

The `gpio_basic_api` test is rather ... basic. Since this is a new driver, could you please ensure `tests/drivers/gpio/gpio_api_1pin` passes as well?

CkovMk commented:
@henrikbrixandersen Confirmed that gpio_api_1pin passed. No need to add overlay or config.

However, I see most test are **SKIPPED** because "Simultaneous pin in/out mode is not supported.' This is true for our hardware. It seems the test is trying to set an LED pin to in and out at the same time and read back the level of the output.

CkovMk commented:
@henrikbrixandersen Confirmed that gpio_api_1pin passed. No need to add overlay or config.

However, I see most test are **SKIPPED** because "Simultaneous pin in/out mode is not supported.' This is true for our hardware. It seems the test is trying to set an LED pin to in and out at the same time and read back the level of the output.

henrikbrixandersen commented:
> @henrikbrixandersen Confirmed that gpio_api_1pin passed. No need to add overlay or config.
> 
> 
> 
> However, I see most test are **SKIPPED** because "Simultaneous pin in/out mode is not supported.' This is true for our hardware. It seems the test is trying to set an LED pin to in and out at the same time and read back the level of the output.

Okay, that's fair. Thank you for confirming.

kartben commented on boards/arm64/mimx93_evk/mimx93_evk_a55.dts:
it would probably make sense to update the "Supported Features" section of the board's README.
